src/wmtk/utils/polygon_mesh_topology_initialization.cpp

```diff
@@ -0,0 +1,409 @@
+    opp_ext.topRows(n_he) = opp;
+    for (long i = 0; i < n_bnd_he; ++i) {
+        opp_ext(bnd_he[i]) = n_he + i;
+        opp_ext(n_he + i) = bnd_he[i];
+    }
+
+    // Add boundary halfedge data for the extended next and to halfedge array
+    VectorXl next_ext(n_he + n_bnd_he);
+    VectorXl to_ext(n_he + n_bnd_he);
+    next_ext.topRows(n_he) = next;
+    to_ext.topRows(n_he) = to;
+    for (long i = 0; i < n_bnd_he; ++i) {
+        long he = bnd_he[i];
+        long he_it = next(he);
+        while (opp(he_it) != -1) {
+            he_it = next(opp(he_it));
+        }
+        next_ext(opp_ext(he_it)) = opp_ext(he);
+        to_ext(opp_ext(he_it)) = to(he);
+    }
+
+    // Build faces for extended halfedge, including new boundary faces
+    std::vector<std::vector<long>> faces = build_orbits(next_ext);
+    long n_f_ext = faces.size();
+
+    // Add boundary halfedge data for the extended he2f array, recording a halfedge per new face
+    std::vector<long> bnd_loops(0);
+    VectorXl he2f_ext(n_he + n_bnd_he);
+    he2f_ext.topRows(n_he) = he2f;
+    long boundary_face_index = -1; // use unique negative indices for boundaries
+    for (long i = 0; i < n_f_ext; ++i) {
+        if (faces[i][0] >= n_he) {
+            bnd_loops.push_back(faces[i][0]);
+            long face_size = faces[i].size();
+            for (long j = 0; j < face_size; ++j) {
+                he2f_ext(faces[i][j]) = boundary_face_index;
+            }
+            --boundary_face_index;
+        }
+    }
+
+    return std::make_tuple(next_ext, opp_ext, to_ext, he2f_ext, bnd_loops);
+}
+
+// Generate halfedge permutation to make opp implicit
+VectorXl build_implicit_opp_reindex(Eigen::Ref<const VectorXl> opp)
+{
+    long n_he = opp.size();
+
+    // Get edge pairs
+    std::vector<std::vector<long>> edges = build_orbits(opp);
+    long n_e = edges.size();
+    assert(2 * n_e == n_he);
+
+    // Build permutation mapping old halfedge indices to new indices with implicit opp
+    VectorXl he_reindex(n_he);
+    for (long e = 0; e < n_e; ++e) {
+        const auto& edge = edges[e];
+        assert(edge.size() == 2);
+        he_reindex(edge[0]) = 2 * e;
+        he_reindex(edge[1]) = 2 * e + 1;
+    }
+
+    return he_reindex;
+}
+
+// Implicit opposite map for halfedges paired as (e) = {2*e, 2*e + 1}
+long implicit_opp(long h)
+{
+    return ((h % 2) == 0) ? (h + 1) : (h - 1);
+}
+
+} // namespace
+
+std::vector<std::vector<long>> build_orbits(Eigen::Ref<const VectorXl> perm)
+{
+    long n_perm = perm.size();
+
+    // Get the maximum value in perm
+    long max_perm = 0;
+    for (long i = 0; i < n_perm; ++i) {
+        max_perm = std::max(perm(i), max_perm);
+    }
+    std::vector<bool> visited(max_perm + 1, false);
+
+    // Build the cycles of the permutation
+    std::vector<std::vector<long>> cycles(0);
+    for (long i = 0; i < max_perm + 1; ++i) {
+        if (!visited[i]) {
+            cycles.push_back(std::vector<long>());
```

Comment:
generally try to use `emplace_back` as a habit - this code will probably elide / get optimized in but in some cases it can create a copy